From 3e0418105cb1c044ed840fcf355441401198a949 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Mon, 18 Apr 2022 13:14:26 +0300
Subject: [PATCH] [SW-66329] efuse: refactor preboot_fuse_read function

The previos implementation assumes notification methods like prints
and psoc status bits are available, which is not the case for all
the CPUs that are using the efuse infrastructure.
In addtion not all CPUs are allowed to run the fuse initialization.
This Patch moves the cpu specific functionality to __weak functions.

Change-Id: I334c90da27164d769657ff6ee8d6d23dc159848e
Signed-off-by: Igal Zeltser <izeltser@habana.ai>
---

diff --git a/lib/common/efuse.c b/lib/common/efuse.c
index a6066b9..6026725 100644
--- a/lib/common/efuse.c
+++ b/lib/common/efuse.c
@@ -28,7 +28,7 @@
 
 static int efuse_init;
 
-int preboot_fuse_init(void)
+__weak int preboot_fuse_init(void)
 {
 	int i, offset;
 	u32 read_reg;
@@ -62,21 +62,31 @@
 	return 0;
 }
 
+__weak void fuse_notify_init_fail(void)
+{
+	cpu_boot_err_set(CPU_BOOT_ERR_EFUSE_FAIL);
+	preboot_puts("Efuse not initialized\n");
+}
+
+__weak void fuse_notify_invalid_offset(void)
+{
+	cpu_boot_err_set(CPU_BOOT_ERR_EFUSE_FAIL);
+	preboot_puts("Invalid efuse word offset\n");
+}
+
 __weak int preboot_fuse_read(u32 word, u32 *val)
 {
 	int offset, ret;
 	u32 read_val;
 
 	if (word >= efuse.words_per_bank) {
-		cpu_boot_err_set(CPU_BOOT_ERR_EFUSE_FAIL);
-		preboot_puts("Invalid efuse word offset\n");
+		fuse_notify_invalid_offset();
 		return -1;
 	}
 
 	ret = preboot_fuse_init();
 	if (ret == -2) {
-		cpu_boot_err_set(CPU_BOOT_ERR_EFUSE_FAIL);
-		preboot_puts("Efuse not initialized\n");
+		fuse_notify_init_fail();
 		return ret;
 	}
 
diff --git a/lib/common/efuse.h b/lib/common/efuse.h
index df78b04..2aa57a6 100644
--- a/lib/common/efuse.h
+++ b/lib/common/efuse.h
@@ -18,5 +18,8 @@
 
 int preboot_fuse_init(void);
 int preboot_fuse_read(u32 word, u32 *val);
+void fuse_notify_init_fail(void);
+void fuse_notify_invalid_offset(void);
+
 
 #endif /* _EFUSE_H_ */
